File: .claude/templates/boundary-doubles/fake-llm.ts
/**
 * Deterministic fake LLM client (boundary-test-doubles kit, gap G34). TypeScript sibling
 * of fake_llm.py: golden structured responses keyed by (operation, stable request hash),
 * so LLM-backed flows are deterministic in tests. The harness assumes a raw-fetch LLM
 * wrapper (no TS Anthropic SDK), so this doubles that wrapper's request payload. Return a
 * response validated against the same Zod / json_schema contract the real wrapper uses.
 */
import { promises as fs } from 'fs';
import * as path from 'path';
import { createHash } from 'crypto';

/** Stable 16-char hash of the request payload, order-independent (mirrors Python's
 * json.dumps(sort_keys=True)) so equivalent payloads map to the same golden. */
export function requestKey(payload: unknown): string {
  const canonical = JSON.stringify(sortValue(payload));
  return createHash('sha256').update(canonical).digest('hex').slice(0, 16);
}

export class GoldenNotFoundError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'GoldenNotFoundError';
  }
}

export class FakeLLMClient {
  constructor(private readonly fixturesRoot: string = 'tests/fixtures/llm') {}

  private pathFor(operation: string, key: string): string {
    return path.join(this.fixturesRoot, operation, `${key}.json`);
  }

  async respond(operation: string, payload: unknown): Promise<unknown> {
    const key = requestKey(payload);
    const p = this.pathFor(operation, key);
    try {
      return JSON.parse(await fs.readFile(p, 'utf8'));
    } catch (err) {
      if ((err as NodeJS.ErrnoException).code === 'ENOENT') {
        throw new GoldenNotFoundError(
          `no golden LLM response for ${operation}/${key} at ${p}; ` +
          `record it once against the real model via recordGolden()`,
        );
      }
      throw err;
    }
  }

  async recordGolden(operation: string, payload: unknown, response: unknown): Promise<string> {
    const key = requestKey(payload);
    const p = this.pathFor(operation, key);
    await fs.mkdir(path.dirname(p), { recursive: true });
    await fs.writeFile(p, JSON.stringify(response, null, 2));
    return p;
  }
}

function sortValue(v: unknown): unknown {
  if (Array.isArray(v)) return v.map(sortValue);
  if (v && typeof v === 'object') {
    const src = v as Record<string, unknown>;
    const out: Record<string, unknown> = {};
    for (const k of Object.keys(src).sort()) out[k] = sortValue(src[k]);
    return out;
  }
  return v;
}

File: .claude/templates/boundary-doubles/fake_llm.py
"""Deterministic fake LLM client (boundary-test-doubles kit, gap G34).

Returns golden structured responses keyed by (operation, stable request hash) so
LLM-backed flows are deterministic in tests. code-gen mandates tool_use/JSON-schema
output, so the golden is a validated JSON object, not free text. Drop-in for the
LLM wrapper's SDK client under HARNESS_TEST_REPLAY=1.

`request_key` is defined below the class deliberately (it is resolved at call
time) to keep each definition small and unambiguous.
"""
import hashlib
import json
from pathlib import Path


class GoldenNotFoundError(RuntimeError):
    """Raised in replay mode when no golden response exists for a request.

    In a forced-replay regression run this signals the flow would have called the
    real model — a hard failure, not a live call.
    """


class FakeLLMClient:
    def __init__(self, fixtures_root: str = "tests/fixtures/llm"):
        self._dir = Path(fixtures_root)

    def _path(self, operation: str, key: str) -> Path:
        return self._dir / operation / f"{key}.json"

    def respond(self, operation: str, payload: dict) -> dict:
        key = request_key(payload)
        p = self._path(operation, key)
        if not p.exists():
            raise GoldenNotFoundError(
                f"no golden LLM response for {operation}/{key} at {p}; "
                f"record it once against the real model via record_golden()"
            )
        return json.loads(p.read_text())

    def record_golden(self, operation: str, payload: dict, response: dict) -> Path:
        key = request_key(payload)
        p = self._path(operation, key)
        p.parent.mkdir(parents=True, exist_ok=True)
        p.write_text(json.dumps(response, indent=2, sort_keys=True))
        return p


def request_key(payload: dict) -> str:
    canonical = json.dumps(payload, sort_keys=True, separators=(",", ":"))
    return hashlib.sha256(canonical.encode("utf-8")).hexdigest()[:16]
